### PR TITLE
Reorganize the Java client to provide multiple implementations

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

We'd like to provide multiple Java client implementations rather than
just one:

- the Armeria-based implementation that uses HTTP API
- the Armeria-based legacy implemenation that uses Thrift (the current impl)
- the lightweight implementation that does not depend on Armeria

Modifications:

- Make `common` and `client` free from Thrift.
  - Add `common-legacy` that depends on `common` and includes
    Thrift-specific classes
- Add `client-armeria` and `client-armeria-legacy`
  - Move Armeria-specific code to `client-armeria`
  - Move Thrift-specific code to `client-armeria-legacy`
  - Make `AbstractWatcher` Armeria-free
  - Rename `DefaultCentralDogma` to `LegacyCentralDogma`
  - Introduce class hierarchy to the client builder
    - `client` - `AbstractCentralDogmaBuilder`
    - `client-armeria` - `AbstractArmeriaCentralDogmaBuilder` extends
      `AbstractCentralDogmaBuilder`
    - `client-armeria-legacy` - `LegacyCentralDogmaBuilder` extends
      `AbstractArmeriaCentralDogmaBuilder`
    - `client-armeria` would have `ArmeriaCentralDogmaBuilder` that
      extends `AbstractArmeriaCentralDogmaBuilder`.
  - As a result, the factory methods such as `forHost()` and `forProfile()`
    have been removed.
- Do not use the `CentralDogmaException` generated by Thrift in the
  public API.
  - Add `common.CentralDogmaException` and its subtypes that replace the
    Thrift exception
  - Move some exceptions in `server` to `common`
  - Split `QueryException` into `QuerySyntaxException` and
    `QueryExecutionException` for clarification
  - Use `StorageException` only when something really unexpected occurred
  - Remove `StorageExistsException` and `StorageNotFoundException` which
    didn't have much role but only complicated type hierarchy

Result:

- We can add more `CentralDogma` implementations in the future.

### CODE REVIEW

File: client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2017 LINE Corporation
```
Comment (imasahiro) [inline client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java]: 2018?

Comment (trustin) [reply client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java]: Fixed :-)

File: client/java/src/main/java/com/linecorp/centraldogma/internal/client/AbstractWatcher.java [hunk 2]
```diff
@@ -49,8 +55,46 @@
     private static final long MAX_INTERVAL_MILLIS = TimeUnit.MINUTES.toMillis(1);
     private static final double JITTER_RATE = 0.2;
     private static final long WATCH_TIMEOUT_MILLIS = TimeUnit.MINUTES.toMillis(1);
-    private static final Supplier<? extends Backoff> BACKOFF_SUPPLIER =
-            () -> Backoff.exponential(MIN_INTERVAL_MILLIS, MAX_INTERVAL_MILLIS).withJitter(JITTER_RATE);
+
+    private static long nextDelayMillis(int numAttemptsSoFar) {
```
Comment (minwoox) [inline client/java/src/main/java/com/linecorp/centraldogma/internal/client/AbstractWatcher.java]: Is there any reason that you implemented this logic by yourself?

File: client/java/src/main/java/com/linecorp/centraldogma/client/AbstractCentralDogmaBuilder.java [hunk 3]
```diff
@@ -271,69 +258,41 @@ public CentralDogmaBuilder profile(ClassLoader classLoader, Iterable<String> pro
...
+        } else {
+            endpoint = InetSocketAddress.createUnresolved(host, port);
+        }
+        return endpoint;
     }
 
     /**
-     * Sets the {@link ClientFactory} that will create an underlying
-     * <a href="https://line.github.io/armeria/">Armeria</a> client which performs the actual socket I/O.
+     * Returne the name of the selected profile.
```
Comment (minwoox) [inline client/java/src/main/java/com/linecorp/centraldogma/client/AbstractCentralDogmaBuilder.java]: Returns?

File: client/java/src/main/java/com/linecorp/centraldogma/internal/client/AbstractWatcher.java (hunk 2 above)
Comment (trustin) [reply client/java/src/main/java/com/linecorp/centraldogma/internal/client/AbstractWatcher.java]: Because it has to be Armeria-free?

Comment (minwoox) [reply client/java/src/main/java/com/linecorp/centraldogma/internal/client/AbstractWatcher.java]: Ah got it!

File: client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    /**
+     * Returns the {@link Endpoint} this client will connect to, derived from {@link #hosts()}.
+     */
+    protected final Endpoint endpoint() {
+        final Set<InetSocketAddress> hosts = hosts();
+        checkState(!hosts.isEmpty(), "no hosts were added.");
+
+        final Endpoint endpoint;
+        if (hosts.size() == 1) {
+            endpoint = toEndpoint(hosts.iterator().next());
```
Comment (minwoox) [inline client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java]: `Iterables.getFirst`? :-)

File: site/src/sphinx/client-java.rst [hunk 5]
```diff
@@ -35,7 +35,7 @@ Maven:
       ...
       <dependency>
         <groupId>com.linecorp.centraldogma</groupId>
-        <artifactId>centraldogma-client</artifactId>
+        <artifactId>centraldogma-client-armeria-legacy</artifactId>
```
Comment (minwoox) [inline site/src/sphinx/client-java.rst]: No need to update the gradle part?

File: client/java-armeria-legacy/src/main/java/com/linecorp/centraldogma/client/armeria/legacy/LegacyCentralDogma.java [hunk 6]
```diff
@@ -404,35 +425,91 @@
...
-            if (e instanceof InvocationTargetException) {
-                cause = MoreObjects.firstNonNull(e.getCause(), e);
-            } else {
-                cause = e;
-            }
-            CompletableFuture<T> failedFuture = new CompletableFuture<>();
-            failedFuture.completeExceptionally(cause);
+            final CompletableFuture<T> failedFuture = new CompletableFuture<>();
+            failedFuture.completeExceptionally(convertCause(e));
             return failedFuture;
```
[the remark this answers is not in the document]
Comment (trustin) [reply client/java-armeria-legacy/src/main/java/com/linecorp/centraldogma/client/armeria/legacy/LegacyCentralDogma.java]: Thanks!

File: client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+ */
+public class AbstractArmeriaCentralDogmaBuilder<B extends AbstractArmeriaCentralDogmaBuilder<B>>
+        extends AbstractCentralDogmaBuilder<B> {
+
+    @VisibleForTesting
+    static final AtomicLong nextAnonymousGroupId = new AtomicLong();
+
+    private ClientFactory clientFactory = ClientFactory.DEFAULT;
+    private ArmeriaClientConfigurator clientConfigurator = cb -> {
+    };
```
[the remark this answers is not in the document]
Comment (trustin) [reply client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java]: OK to change it. Could you send a PR?

[the remark this answers is not in the document]
Comment (minwoox) [reply client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java]: I will. Thank you~!

File: site/src/sphinx/client-java.rst (hunk 5 above)
Comment (trustin) [reply site/src/sphinx/client-java.rst]: Oops. Fixed.

File: client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java (hunk 4 above)
Comment (trustin) [reply client/java-armeria/src/main/java/com/linecorp/centraldogma/client/armeria/AbstractArmeriaCentralDogmaBuilder.java]: Yeah :-)

File: client/java/src/main/java/com/linecorp/centraldogma/client/AbstractCentralDogmaBuilder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (trustin) [reply client/java/src/main/java/com/linecorp/centraldogma/client/AbstractCentralDogmaBuilder.java]: Fixed

**Comment by trustin** [comment]
Thanks, reviewers!

**Pull request merged**